src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs

```diff
@@ -154,5 +155,136 @@ public static void VerifyExpiration_LocalTime(DateTime verificationTime, bool sh
                 }
             }
         }
+
+        [Fact]
+        public static void BuildChain_WithApplicationPolicy_Match()
+        {
+            using (var msCer = new X509Certificate2(TestData.MsCertificate))
+            {
+                X509Chain chain = new X509Chain();
+                // Code Signing
+                chain.ChainPolicy.ApplicationPolicy.Add(new Oid("1.3.6.1.5.5.7.3.3"));
+                chain.ChainPolicy.VerificationTime = msCer.NotBefore.AddHours(2);
+                chain.ChainPolicy.VerificationFlags =
+                    X509VerificationFlags.AllowUnknownCertificateAuthority;
+
+                chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+
+                bool valid = chain.Build(msCer);
+                Assert.True(valid, "Chain built validly");
+            }
+        }
+
+        [Fact]
+        public static void BuildChain_WithApplicationPolicy_NoMatch()
+        {
+            using (var cert = new X509Certificate2(TestData.MsCertificate))
+            {
+                X509Chain chain = new X509Chain();
+                // Gibberish.  (Code Signing + ".1")
+                chain.ChainPolicy.ApplicationPolicy.Add(new Oid("1.3.6.1.5.5.7.3.3.1"));
+                chain.ChainPolicy.VerificationFlags =
+                    X509VerificationFlags.AllowUnknownCertificateAuthority;
+
+                chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                chain.ChainPolicy.VerificationTime = cert.NotBefore.AddHours(2);
+
+                bool valid = chain.Build(cert);
+                Assert.False(valid, "Chain built validly");
+
+                Assert.True(chain.ChainElements.Count > 0, "chain.ChainElements.Count > 0");
+
+                Assert.NotSame(cert, chain.ChainElements[0].Certificate);
+                Assert.Equal(cert, chain.ChainElements[0].Certificate);
+
+                X509ChainStatus[] chainElementStatus = chain.ChainElements[0].ChainElementStatus;
+                Assert.True(chainElementStatus.Length >= 1, "chainElementStatus.Length >= 1");
+
+                bool foundExpectedFlag = false;
+
+                foreach (X509ChainStatus statusItem in chainElementStatus)
+                {
+                    if (statusItem.Status == X509ChainStatusFlags.NotValidForUsage)
+                    {
+                        foundExpectedFlag = true;
+                        break;
+                    }
+                }
+                
+                Assert.True(foundExpectedFlag, "chain.ChainElements[0].ChainElementStatus contains NotValidForUsage");
```

Comment:
Yeah, that one isn't pretty, but it seems to be the most compact.

(Intentional failure)
```
     System.Security.Cryptography.X509Certificates.Tests.ChainTests.BuildChain_
  WithCertificatePolicy_NoMatch [FAIL]
        Assert.Contains() Failure
        Not found: (filter expression)
        In value:  X509ChainStatus[] [System.Security.Cryptography.X509Certific
  ates.X509ChainStatus, System.Security.Cryptography.X509Certificates.X509Chain
  Status]
```